File: src/test/java/org/shsts/tinactory/datagen/content/builder/VeinBuilder.java
package org.shsts.tinactory.datagen.content.builder;

import javax.annotation.ParametersAreNonnullByDefault;
import net.minecraft.MethodsReturnNonnullByDefault;
import net.minecraft.util.Unit;
import org.shsts.tinactory.content.material.MaterialSet;
import org.shsts.tinactory.content.material.OreVariant;
import org.shsts.tinactory.content.recipe.OreAnalyzerRecipe;
import org.shsts.tinactory.core.builder.Builder;
import org.shsts.tinactory.datagen.builder.TechBuilder;

import java.util.ArrayList;
import java.util.List;

import static org.shsts.tinactory.content.AllRecipes.ORE_ANALYZER;
import static org.shsts.tinactory.datagen.builder.TechBuilder.RANK_PER_VOLTAGE;
import static org.shsts.tinactory.datagen.content.Technologies.BASE_ORE;
import static org.shsts.tinactory.datagen.content.Technologies.TECHS;
import static org.shsts.tinactory.test.TinactoryTest.DATA_GEN;

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
public class VeinBuilder<P> extends Builder<Unit, P, VeinBuilder<P>> {
    public static final int VEIN_TECH_RANK = RANK_PER_VOLTAGE / 2;

    private final String id;
    private final int rank;
    private final double rate;
    private final OreAnalyzerRecipe.Builder builder;
    private final List<MaterialSet> ores = new ArrayList<>();
    private boolean baseOre = false;
    private boolean primitive = false;
    private OreVariant variant = null;

    private VeinBuilder(P parent, String id, int rank, double rate) {
        super(parent);
        this.id = id;
        this.rank = rank;
        this.rate = rate;
        this.builder = ORE_ANALYZER.recipe(DATA_GEN, id).rate(rate);
    }

    public static <P> VeinBuilder<P> factory(P parent, String id, int rank, double rate) {
        var builder = new VeinBuilder<>(parent, id, rank, rate);
        return builder.onBuild(builder::onRegister);
    }

    public VeinBuilder<P> ore(MaterialSet material, double rate) {
        builder.outputItem(material.entry("raw"), 1, rate);
        ores.add(material);
        if (variant == null) {
            variant = material.oreVariant();
            builder.inputItem(() -> variant.baseItem, 1);
        }
        assert variant == material.oreVariant();
        return this;
    }

    public VeinBuilder<P> primitive() {
        builder.primitive();
        primitive = true;
        baseOre = true;
        return this;
    }

    public VeinBuilder<P> base() {
        baseOre = true;
        return this;
    }

    @Override
    protected Unit createObject() {
        return Unit.INSTANCE;
    }

    private void onRegister() {
        assert variant != null;
        assert rate > 0d;
        assert !ores.isEmpty();
        if (!primitive) {
            builder.voltage(variant.voltage);
        }
        var tech = BASE_ORE.get(variant);
        var baseProgress = 30L;
        if (!baseOre) {
            tech = TECHS.builder("ore/" + id, TechBuilder::factory)
                .maxProgress(baseProgress)
                .displayItem(ores.get(0).loc("raw"))
                .depends(tech)
                .rank(rank + VEIN_TECH_RANK + 1)
                .researchVoltage(variant.voltage)
                .register();
        }

        if (!primitive) {
            builder.requireTech(tech);
        }
        builder.build();
    }
}

File: src/test/kotlin/org/shsts/tinactory/datagen/content/builder/VeinBuilder.kt
package org.shsts.tinactory.datagen.content.builder

import org.shsts.tinactory.Tinactory.REGISTRATE
import org.shsts.tinactory.content.AllMaterials.getMaterial
import org.shsts.tinactory.content.recipe.OreAnalyzerRecipe
import org.shsts.tinactory.core.electric.Voltage
import org.shsts.tinactory.core.material.MaterialSet
import org.shsts.tinactory.core.material.OreVariant
import org.shsts.tinactory.datagen.builder.TechBuilder
import org.shsts.tinactory.datagen.builder.TechBuilder.RANK_PER_VOLTAGE
import org.shsts.tinactory.datagen.content.Technologies.BASE_ORE
import org.shsts.tinactory.datagen.content.Technologies.TECHS
import org.shsts.tinactory.test.TinactoryTest.DATA_GEN

class VeinBuilder(private val id: String, private val rank: Int, private val rate: Double) {
    var primitive = false
    var baseOre = false
    private var variant: OreVariant? = null
    private var block: ProcessingRecipeBuilder<OreAnalyzerRecipe.Builder>.() -> Unit = {}
    private val ores = mutableListOf<MaterialSet>()

    companion object {
        const val VEIN_TECH_RANK = RANK_PER_VOLTAGE / 2
    }

    private fun chain(another: ProcessingRecipeBuilder<OreAnalyzerRecipe.Builder>.() -> Unit) {
        val oldBlock = block
        block = {
            oldBlock()
            another()
        }
    }

    fun variant(value: OreVariant) {
        variant = value
        chain {
            input(value.baseItem)
        }
    }

    fun ore(name: String, rate: Double) {
        val mat = getMaterial(name)
        if (variant == null) {
            variant(mat.oreVariant())
        }
        chain {
            if (mat.hasItem("raw")) {
                output(mat, "raw", rate = rate)
            } else {
                output(mat, "raw_fluid", rate = rate)
            }
        }
        ores.add(mat)
    }

    fun build() {
        val variant1 = variant!!
        assert(rate > 0)
        assert(ores.isNotEmpty())
        val id1 = "${variant1.serializedName}/$id"

        val baseTech = BASE_ORE.getValue(variant1)
        val tech = if (baseOre || primitive) {
            baseTech
        } else {
            TECHS.builder("ore/$id1") { handler, parent, loc ->
                TechBuilder.factory(handler, parent, loc)
            }.run {
                maxProgress(30)
                val mat = ores[0]
                if (mat.hasItem("raw")) {
                    displayItem(mat.item("raw"))
                } else {
                    displayItem(mat.item("raw_fluid"))
                }
                depends(baseTech)
                researchVoltage(variant1.voltage)
                rank(rank + 1 + VEIN_TECH_RANK)
                register()
            }
        }

        val builder = run {
            val recipeType = REGISTRATE.getRecipeType<OreAnalyzerRecipe.Builder>("ore_analyzer")
            val builder = recipeType.recipe(DATA_GEN, id1).apply {
                rate(this@VeinBuilder.rate)
            }
            ProcessingRecipeBuilder(builder).apply {
                simpleDefaults()
                amperage = 0.125
                workTicks(128)
            }
        }

        builder.apply(block)
        if (primitive) {
            builder.voltage(Voltage.PRIMITIVE)
        } else {
            builder.voltage(variant1.voltage)
            builder.extra {
                requireTech(tech)
            }
        }
        builder.build()
    }
}
